Mathlib/MeasureTheory/Function/JacobianOneDim.lean

```diff
@@ -60,6 +66,319 @@ theorem integral_image_eq_integral_abs_deriv_smul
+          apply le_antisymm  _ (hf hx.1.1 hy.1 hy.2.1.le)
+          rw [← fpx]
+          exact hf hy.1 ps₁.1 hy.2.2.le
+        have : HasDerivWithinAt (fun y ↦ f x) 0 (s ∩ Ioo x p) x :=
+          hasDerivWithinAt_const x (s ∩ Ioo x p) (f x)
+        exact this.congr E rfl
+      have K' : HasDerivWithinAt f (f' x) (s ∩ Ioo x p) x :=
+        (hf' x hx.1.1).mono inter_subset_left
+      apply UniqueDiffWithinAt.eq_deriv _ _ K' K
+      have J1 : (s ∩ Ioo x p) \ {x} = (s ∩ Ioo x p) := by simp
+      have J2 : 𝓝[s ∩ Ioo x p] x = 𝓝[s ∩ Ioi x] x := by
+        simp [nhdsWithin_inter, nhdsWithin_Ioo_eq_nhdsGT px]
+      rw [uniqueDiffWithinAt_iff_accPt, accPt_principal_iff_nhdsWithin, J1, J2]
+      simp only [mem_inter_iff, mem_diff, hx.1.1, mem_union, mem_setOf_eq, true_and, not_or,
+        mem_preimage, b, s₁, a] at hx
+      exact neBot_iff.2 hx.1.1
+  · /- We have to show that the derivative is nonnegative at points of `c`. As these points are
+    not isolated in `s`, this follows from the fact that `f` is monotone on `s`. -/
+    intro x hx
+    apply (hf' x hx.1.1).nonneg_of_monotoneOn _ hf
+    simp only [mem_diff, hx.1.1, mem_union, mem_setOf_eq, true_and, not_or, c, s₁, a, b] at hx
+    rw [accPt_principal_iff_nhdsWithin]
+    have : (𝓝[s ∩ Iio x] x).NeBot := neBot_iff.2 hx.1.2
+    apply this.mono
+    apply nhdsWithin_mono
+    rintro y ⟨yt, yx : y < x⟩
+    exact ⟨yt, by simpa using yx.ne⟩
+  · intro x hx y hy hxy
+    contrapose! hxy
+    wlog H : x < y generalizing x y with h
+    · have : y < x := by order
+      exact (h hy hx hxy.symm this).symm
+    refine fun h ↦ hx.2 ⟨hx.1, ?_⟩
+    exact ⟨x, y, hx.1, hy.1, H, rfl, h.symm⟩
+
+/- Change of variable formula for differentiable functions: if a real function `f` is
+monotone and differentiable on a measurable set `s`, then the Lebesgue integral of a function
+`u : ℝ → ℝ≥0∞` on `f '' s` coincides with the integral of `(f' x) * u ∘ f` on `s`.
+Note that the measurability of `f '' s` is given by `MeasurableSet.image_of_monotoneOn`. -/
+theorem lintegral_image_eq_lintegral_deriv_mul_of_monotoneOn (hs : MeasurableSet s)
+    (hf' : ∀ x ∈ s, HasDerivWithinAt f (f' x) s x) (hf : MonotoneOn f s) (u : ℝ → ℝ≥0∞) :
+    ∫⁻ x in f '' s, u x = ∫⁻ x in s, ENNReal.ofReal (f' x) * u (f x) := by
+  rcases exists_decomposition_of_monotoneOn_hasDerivWithinAt hs hf hf' with
+    ⟨a, b, c, h_union, ha, hb, hc, h_disj, h_disj', a_count, fb_count, deriv_b, deriv_c, inj_c⟩
+  have I : ∫⁻ x in s, ENNReal.ofReal (f' x) * u (f x)
+      = ∫⁻ x in c, ENNReal.ofReal (f' x) * u (f x) := by
+    have : ∫⁻ x in a, ENNReal.ofReal (f' x) * u (f x) = 0 := by
+      have : volume a = 0 := a_count.measure_zero volume
+      exact setLIntegral_measure_zero a _ this
+    rw [← h_union, lintegral_union (hb.union hc) h_disj, this, zero_add]
+    have : ∫⁻ x in b, ENNReal.ofReal (f' x) * u (f x) = 0 :=
+      setLIntegral_eq_zero hb (fun x hx ↦ by simp [deriv_b x hx])
```

Comment:
You could inline this have --- feel free to if you think this is readable enough.

Reply:
Sure, that's something we could disagree on. (And perhaps I'd even agree with you. When reading this proof for the first time, as opposed to scrutinising it, I certainly see the appeal of reading the have in the source code.)